src/sentry/monitors/processing_errors.py

```diff
@@ -36,3 +54,127 @@ class ProcessingErrorType(Enum):
+        return {
+            "type": self.type.name,
+            "data": self.data,
+        }
+
+    @classmethod
+    def from_dict(cls, processing_error_data: ProcessingErrorData) -> ProcessingError:
+        return cls(
+            ProcessingErrorType[processing_error_data["type"]],
+            processing_error_data["data"],
+        )
+
+
+class CheckinProcessingErrorData(TypedDict):
+    errors: list[ProcessingErrorData]
+    checkin: CheckinItemData
+
+
+@dataclasses.dataclass(frozen=True)
+class CheckinProcessingError:
+    errors: list[ProcessingError]
+    checkin: CheckinItem
+
+    def to_dict(self) -> CheckinProcessingErrorData:
+        return {
+            "errors": [error.to_dict() for error in self.errors],
+            "checkin": self.checkin.to_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, data: CheckinProcessingErrorData) -> CheckinProcessingError:
+        return cls(
+            errors=[ProcessingError.from_dict(error) for error in data["errors"]],
+            checkin=CheckinItem.from_dict(data["checkin"]),
+        )
+
+
+class CheckinProcessErrorsManager:
+    def _get_cluster(self):
+        return redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    def store(self, error: CheckinProcessingError, monitor: Monitor | None):
+        if monitor is None:
+            # Attempt to get the monitor from the checkin info if we failed to retrieve it during ingestion
+            try:
+                monitor = Monitor.objects.get(
+                    project_id=error.checkin.message["project_id"],
+                    slug=error.checkin.payload["monitor_slug"],
+                )
+            except Monitor.DoesNotExist:
+                pass
+        if monitor:
+            error_identifier = f"monitor:{monitor.id}"
+        else:
+            error_identifier = f'project:{error.checkin.message["project_id"]}'
+
+        error_key = f"monitors.processing_errors.{error_identifier}"
+        serialized_error = json.dumps(error.to_dict())
+        redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+        pipeline = redis_client.pipeline(transaction=False)
+        pipeline.zadd(error_key, {serialized_error: error.checkin.ts.timestamp()})
+        # Cap the error list to the `MAX_ERRORS_PER_SET` most recent errors
+        pipeline.zremrangebyrank(error_key, 0, -(MAX_ERRORS_PER_SET + 1))
+        pipeline.expire(error_key, MONITOR_ERRORS_LIFETIME)
+        pipeline.execute()
+
+    def build_monitor_identifier(self, monitor: Monitor) -> str:
+        return f"monitor:{monitor.id}"
+
+    def get_for_monitor(self, monitor: Monitor) -> list[CheckinProcessingError]:
+        return self._get_for_entity(self.build_monitor_identifier(monitor))
+
+    def build_project_identifier(self, project: Project) -> str:
+        return f"project:{project.id}"
+
+    def get_for_project(self, project: Project) -> list[CheckinProcessingError]:
```

Comment:
```suggestion
    def get_for_project(self, project: Project) -> Sequence[CheckinProcessingError]:
```

Reply:
Not that clear on what is preferred here. We know we're returning an actual list, isn't it better to be specific? I see both options in the codebase

Reply:
You're right https://typing.readthedocs.io/en/latest/source/best_practices.html#arguments-and-return-types

> For return values, prefer concrete types (list, dict, etc.) for concrete implementations. The return values of protocols and abstract base classes must be judged on a case-by-case basis.